More GPIO control through CLI

SamuelYvon opened the pull request:
# What's new

- Add a `gpio` command in the CLI to perform various GPIO operations (not limited to write)
- This is related to feature request https://github.com/flipperdevices/flipperzero-firmware/issues/1240

There are a few changes I'm expecting to maybe make before this is merged in:
- I used a goto to cleanup the string_t at the end of the function. However, I have noticed that nowhere else in the furi codebase this pattern is used (except the stmlibs). Is this against your coding guidelines?
- I left the previously available `gpio_set` for "retrocompatibility" but I am not sure this matters here? I could remove it.
- I would have liked to reset all the pins to input mode when the CLI session is finished. Is there a way to hook up to some event? I was not able to find anything.

# Verification 

- Using the interface for GPIO read / write / mode set should be enough

# Checklist (For Reviewer)

- [ ] PR has description of feature/bug or link to Confluence/Jira task
- [ ] Description contains actions to verify feature/bugfix
- [ ] I've built this code, uploaded it to the device and verified feature/bugfix

SamuelYvon committed:
Add more options

skotopes commented on applications/cli/cli_commands.c:
```diff
@@ -248,43 +248,195 @@ void cli_command_led(Cli* cli, string_t args, void* context) {
+    }
+
+// TODO: no other goto in the code base
+// except in /lib. Is this not allowed?
```
Not allowed.

SamuelYvon replied:
OK, I'll reimplement w/o gotos.

skotopes commented on applications/cli/cli_commands.c:
```diff
@@ -248,43 +248,195 @@ void cli_command_led(Cli* cli, string_t args, void* context) {
+
+    if(pin_action == mode) {
+        if(value == 0) { // output
+            LL_GPIO_SetPinMode(gpio[num].port, gpio[num].pin, LL_GPIO_MODE_OUTPUT);
```
No need in LL here, use furi_hal_gpio + furi_hal_resources

skotopes commented on applications/cli/cli_commands.c:
```diff
@@ -248,43 +248,195 @@ void cli_command_led(Cli* cli, string_t args, void* context) {
+
+void cli_command_gpio_ctl(Cli* cli, string_t args, void* context) {
+    UNUSED(context);
+    typedef enum GpioCtlAction_Enum {
```
That's not how we do enums

skotopes commented on applications/cli/cli_commands.c:
```diff
@@ -248,43 +248,195 @@ void cli_command_led(Cli* cli, string_t args, void* context) {
+        pin_action = read;
+    } else {
+        printf("Wrong 2nd argument (%s). Use \"mode\" to change the pin mode, \"set\" to set the value of an output pin or \"read\" to read the value of an input pin.", string_get_cstr(action));
+        goto cli_command_gpio_ctl_cleanup;
```
We don't do goto

SamuelYvon commented:
@skotopes , I'll reimplement w/o gotos. Regarding the existing `gpio-set` command, would you rather I remove it, or implement all three options as different commands (ie, gpio-set, gpio-mode, gpio-read). This would simplify the parsing for each command as well. I also noted your comment for `LL_` for the GPIO access. This is also how `gpio-set` is currently implemented. I'll refactor as part of the next iteration of this PR.

SamuelYvon committed:
Change from one big command to three small commands

These changes follow from the review. Instead of having to
rely in GOTOs (which are forbidden :<( ) to ensure cleaning,
I divided in smaller pieces. Since I had no feedback w.r.t
the previously existing `gpio_set`, this fits more with the
current implementation.

I also tried to interpret "That's not how we do enums" as
to not include a `enum EnumName` decl. when using typedefs.

I also distanced the implementation from using LL_GPIO
in favour of functions provided in `<furi_hal_gpio.h>`
but there was nothing to get the mode, so I stuck with
LL_GPIO for mode check.

SamuelYvon commented:
Please review the latest changes.

These changes follow from the review. Instead of having to
rely in GOTO to ensure cleaning, I divided in smaller pieces. Since I had no feedback w.r.t
the previously existing `gpio_set`, this fits more with the
current implementation.

I also tried to interpret "That's not how we do enums" as
to not include a `enum EnumName` decl. when using typedefs.

I also distanced the implementation from using LL_GPIO
in favour of functions provided in `<furi_hal_gpio.h>`
but there was nothing to get the mode, so I stuck with
LL_GPIO for mode check.

skotopes committed:
Merge branch 'dev' into gpio-cli

skotopes commented:
Hi @SamuelYvon, sorry for delay. I'll approve this PR, but will rewrite it's code a little bit to match our standard cli command routine
